### PR TITLE
Log Factory and Reworking tablet manager setup steps for the same

### PR DESCRIPTION
**Pull request by anirbanr-fb**
Summary:
1. Creating a log factory class which can be passed in server options
 The passed in log factory will be used to create the log implementation

2. Reworking how setup is done in the tablet server so that we can
create the log and consensus before we actually start consensus.

Test Plan: Tested this in shadow setups using MySQL replication

Reviewers: iRitwik

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: src/kudu/tserver/tserver_admin.proto
Comment (iRitwik) [inline src/kudu/tserver/tserver_admin.proto:33]: Do we need to differentiate between initializing, initialized and running states?

File: src/kudu/tserver/tablet_server_options.h
Comment (iRitwik) [inline src/kudu/tserver/tablet_server_options.h:44]: I thought `tablet_server_options` was for primitive setup variables for tserver.

File: src/kudu/tserver/simple_tablet_manager.cc
Comment (iRitwik) [inline src/kudu/tserver/simple_tablet_manager.cc:361]: Left over debug comments?

File: src/kudu/tserver/simple_tablet_manager.cc
Comment (iRitwik) [inline src/kudu/tserver/simple_tablet_manager.cc:280]: Looks like the new status is more for posterity. Ignore my comment on the proto file.

File: src/kudu/tserver/simple_tablet_manager.cc
Comment (iRitwik) [inline src/kudu/tserver/simple_tablet_manager.cc:370]: Where is log factory initialized in the flow? Is there a diff that is going to come on top of this stack?

File: src/kudu/consensus/log.cc
Comment (iRitwik) [inline src/kudu/consensus/log.cc:462]: When would `log_factory` not be initialized?

File: src/kudu/consensus/log.cc
Comment (iRitwik) [inline src/kudu/consensus/log.cc:512]: pointer to a `scoped_refptr` as an out param?

**Review by iRitwik — COMMENTED** [review]
A few questions before I approve.

**Review by iRitwik — APPROVED** [review]
Approving to unblock. We might have to revisit this after handing over leader-election control to plugin in the future.

File: src/kudu/tserver/tserver_admin.proto
Comment (anirbanr-fb) [reply src/kudu/tserver/tserver_admin.proto:33]: Initially the manager is constructed but not usable. INITIALIZED is the
point where we have built a set of datastructures and RUNNING is when raft is running.
Having these states help us put nice asserts and throw out RPC calls if the state is lower than what it can handle.

File: src/kudu/tserver/simple_tablet_manager.cc
Comment (anirbanr-fb) [reply src/kudu/tserver/simple_tablet_manager.cc:361]: I guess left them over from original kudu, but agree that we need to stop worrying about upstream in these files and just remove these comments.

File: src/kudu/tserver/simple_tablet_manager.cc
Comment (anirbanr-fb) [reply src/kudu/tserver/simple_tablet_manager.cc:370]: log_factory is either empty or passed in by application, in our case MySQL plugin. if the plugin is uninitialized it falls back to default Log::Log()

File: src/kudu/consensus/log.cc
Comment (anirbanr-fb) [reply src/kudu/consensus/log.cc:462]: when tablet server options has not used a factory class, for example a future application which just wants to use kuduraft log AS-IS

File: src/kudu/consensus/log.cc
Comment (anirbanr-fb) [reply src/kudu/consensus/log.cc:512]: Yes, this is standard code pattern. All out parameters are pointers. Typical return value continues to remain a Status.

**Comment by anirbanr-fb** [comment]
Replied to some of your comments.

File: src/kudu/tserver/tserver_admin.proto
Comment (anirbanr-fb) [reply src/kudu/tserver/tserver_admin.proto:33]: I don't think we need too fine a distinction, i.e. INITIALIZING and INITIALIZED can be merged together.

**Pull request merged**
